stactask/utils.py

```diff
@@ -44,3 +50,63 @@ def find_collection(
         ),
         None,
     )
+
+
+def _find_collection_from_matchers(
+    collection_matchers: list[dict[str, Any]], item: dict[str, Any]
+) -> Optional[str]:
+    """Find the collection for a given STAC Item represented as a dictionary from a
+    list of collection matcher dictionaries.
+
+    Args:
+        collection_matchers: List of matcher dictionaries with 'type',
+            'collection_name', and 'pattern' (except the 'catch_all' matcher) fields.
+        item: A STAC Item represented as a dictionary.
+
+    Returns:
+        Optional[str]: Returns None if no matcher matches, returns a collection
+        name if one does.
+    """
+    for matcher in collection_matchers:
+        matcher_type = matcher["type"]
+        collection_name: str = matcher["collection_name"]
+
+        if matcher_type == MatcherType.JSONPATH:
+            if stac_jsonpath_match(item, matcher["pattern"]):
+                return collection_name
+        elif matcher_type == MatcherType.CATCH_ALL:
+            return collection_name
+        else:
+            raise ValueError(f"Unknown matcher type: {matcher_type}")
+
+    return None
```

Comment:
This is just a comment, but I really like that this "validation" runs lazily on access. In some cases we really want to run validation early to fail fast (e.g., structural issues, required values...). But this is a good case where we absolutely do not want to throw an error just because a payload includes a matcher we don't recognize.

Why? What if we upgrade a stac-task to a version with support for a new matcher type, but we haven't yet upgraded cirrus or other tasks in a workflow that won't be assigning collections? Trying to use a new matcher type would result in failures if we fail fast, but not if we don't, meaning such validation would be "interoperably not backward-compatible", to make up a hopefully-descriptive phrase.

Reply:
I"m not 100% clear on this, but I take the point that this is a general comment and not a request for a change. 🙂